packages/medusa/src/services/auth.ts

```diff
@@ -1,22 +1,29 @@
 import Scrypt from "scrypt-kdf"
-import { BaseService } from "medusa-interfaces"
+import { BaseService } from "../interfaces"
 import { AuthenticateResult } from "../types/auth"
 import { User } from "../models/user"
 import { Customer } from "../models/customer"
+import UserService from "./user"
+import CustomerService from "./customer"
+
+type InjectedDependencies = {
+  userService: UserService
+  customerService: CustomerService
+}
 
 /**
  * Can authenticate a user based on email password combination
  * @extends BaseService
  */
-class AuthService extends BaseService {
-  constructor({ userService, customerService }) {
-    super()
+class AuthService extends BaseService<AuthService> {
+  protected readonly userService_: UserService
+  protected readonly customerService_: CustomerService
 
-    /** @private @const {UserService} */
-    this.userService_ = userService
+  constructor(cradle: InjectedDependencies) {
```

Comment:
thought: maybe we should reconsider this approach of using the entire cradle: https://github.com/jeffijoe/awilix#:~:text=PROXY%20technically%20allows%20you%20to%20defer%20pulling%20dependencies%20(for%20circular%20dependency%20support)%2C%20but%20this%20isn%27t%20recommended.

Would love to hear your thoughts on this?

Reply:
I would rather prefer an injection such as:

```ts
class InjectedClass {}

class Foo {
    constructor(protected readonly injectedClass: InjectedClass) {}
}
```

It is more verbose but allow to inject only what needed, like other language :)
Was it what you thought?

Reply:
I think I would still prefer `constructor({ injectedService, ... })` as we will need to be able to use the entire cradle in services like `PaymentProviderService, FulfillmentProviderService, TaxProviderService...` + it makes testing a lot easier.

But as far as I understand it from the link above `constructor(cradle)` is not recommended

Reply:
Yeah I see ahah so its better for the ease of use in some cases.
I did not understand that it is not recommended, plus the services are mainly SINGLETON and resolved from the container cache which means that the resolution has already been done.

At the moment, it is already in proxy mode [you can look here](https://github.com/jeffijoe/awilix/blob/3f5c4817baacd99bee7d4c9de9ec8533785a050e/src/container.ts#L233) that is why when you inspect the cradle its a proxy map

Reply:
And for testing both approaches are fine since you can spread it in the constructor if it is in classic mode instead if passing the object 😅

Reply:
wdyt?

Reply:
Because when you destructure in the constructor, `resolve` is still running.

Reply:
@srindom @jeffijoe Classic mode is what I would go for.

Concerning the resolution fail success, from my point of view here is the thing:

If you destructure the cradle it means that the resolution must be done before the instanciation of the target class, but if one of the injected class depend on that class then it fails.

When you access the dependency like this `cradle.classA` you are actually just calling the `get` method of the proxyfied property of the cradle. In other terms, the resolution is done when accessing the property.

I had to dig into that when I did the extender in order to be able to manipulate as expected.

Reply:
@adrien2p yes but destructuring and accessing a property should both be calling `get` on the proxy though.

Example transpilation via TS playground:

<img width="755" alt="image" src="https://user-images.githubusercontent.com/2505474/163417211-f928730f-3c17-4df0-85f6-c6298068e70d.png">

Reply:
> @adrien2p yes but destructuring and accessing a property should both be calling `get` on the proxy though.
> 
> Example transpilation via TS playground:
> 
> <img alt="image" width="755" src="https://user-images.githubusercontent.com/2505474/163417211-f928730f-3c17-4df0-85f6-c6298068e70d.png">

Yes for destructuring I am agree but I was talking about destructuring vs accessing
```
// Destructuring
var Test = /** @class */ (function () {
    function Test(_a) {
        var a = _a.a, b = _a.b;
        this.a = a;
        this.b = b;
    }
    return Test;
}());
// accessing
var Test2 = /** @class */ (function () {
    function Test2(cradle) {
        this.a = cradle.a;
        this.b = cradle.b;
    }
    return Test2;
}());
```

The destructuring does not allow to defer the resolution while the accessing yes.
Why, because you can create a class by passing a temporary object before populating the properties later on. This is the strategy for automatic DI circular dependency. Which is also why in a normal context you should avoid accessing the dependency to manipulate them into the constructor. i.e angular, nestjs for example

Reply:
Yup you don't want to end up using the cradle as a service locator, that'll negate a lot of the benefits of DI regarding reasonability. Looking at a constructor and knowing what a module depends on based on that is very helpful, and if your constructor has 45 dependencies then maybe the design should be rethought 😉  I have been in that situation myself many times before haha. To be honest I actually prefer wiring up dependencies manually instead of using a container these days.

Reply:
@jeffijoe completely agree with you, which is why I suggested the following

```ts
class InjectedClass {}

class Foo {
    constructor(protected readonly injectedClass: InjectedClass) {}
}
````

Also, I quite agree with pure dependency but I think a DI system is quite handy specially when you want to scope your instances or make them transient. In the end it leads you to build your own DI system.
In the end, we should not forget that those principles are here to help testing to be easier.

Reply:
@jeffijoe rho man, I just realised that you have made awilix 😂 I had to build my own DI system years ago and worked around the nestjs one 😂

Reply:
I am only commenting here because I was summoned 😂 

Yes it definitely makes testing easier. Regarding lifetime management though, I feel that can get lost in the container magic. Often people won't know whether a component is scoped, singleton or transient, and if it matters, then why not express that in the code through a factory function? This also makes in-memory caching easier to reason about.

Reply:
@jeffijoe ahah no worries I am glad to discuss that with you.
Agree that the scoping should be explicit, which is why if you look at the extender I did add that option into my decorator in order to see which scope it is using when you look at the class :)

Reply:
> The border between the service locator and DI is thin

There is a pretty huge difference between passing in a "god object" (the locator) that you can take whatever you want from, and defining the minimal requirements for creating an instance of a component. 😉 

For example, I have started defining smaller interfaces so instead of depending on a `IFooService`, I'll depend on a `IFooReader` or `ICanGetFooById` depending on the context. That makes testing *even easier* since now you only have to build fakes for the methods you use. Interface segregation ftw!

Reply:
@jeffijoe also, making test easier as not been applied in medusa at the moment. When I look at the tests, the dependencies are build locally and then passed to the service constructor.
 where I would have created a container at the beginning with the different dependencies and then just override some dependencies (mock for example) and re resolve the container in those cases

Reply:
@jeffijoe I see that we are both SOLID man 😂